initial kubemodel proposal

aaliomer-ibm opened the pull request:
Initial model for kubemodel.

The model uses metered fields that allows for merging kubemodelsets.
The model is improved adding standardized fields and units for type safety and self documentation

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
+//nolint:stylecheck // generated code and package conventions
```
why are you disabling the style linter on all the files?

aaliomer-ibm replied:
will fix it. was giving issues with cldy before moving it

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
-	Name     string    `json:"name"`     // @bingen:field[version=1]
-	Start    time.Time `json:"start"`    // @bingen:field[version=1]
-	End      time.Time `json:"end"`      // @bingen:field[version=1]
+	UID          string   `json:"uid"`          // @bingen:field[version=1]
```
were we going to turn this into a uuid.UUID type?

aaliomer-ibm replied:
Yes! forgot about that but will get on it

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
-	Account  string    `json:"account"`  // @bingen:field[version=1]
-	Name     string    `json:"name"`     // @bingen:field[version=1]
-	Start    time.Time `json:"start"`    // @bingen:field[version=1]
-	End      time.Time `json:"end"`      // @bingen:field[version=1]
```
did we want to keep start/end here?

aaliomer-ibm replied:
will put it back just in case

aaliomer-ibm replied:
actually probably not belong here.

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -1,23 +1,105 @@
-	PodUID                              string    `json:"podUid"`                              // @bingen:field[version=1]
-	Name                                string    `json:"name"`                                // @bingen:field[version=1]
-	Start                               time.Time `json:"start"`                               // @bingen:field[version=1]
-	End                                 time.Time `json:"end"`                                 // @bingen:field[version=1]
```
I think we do still want start/end here

Sean-Holcomb commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -1,25 +1,93 @@
+	if !exists || n.DurationSeconds == 0 {
+		return 0
+	}
+	return KiBToBytes(volume.UsageKiBSeconds) / n.DurationSeconds
```
If we are worried about uint overflow, maybe we should divide by seconds first than convert to bytes

ameijer replied:
I agree with sean here

Sean-Holcomb commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -1,25 +1,93 @@
-	CpuMillicoreUsageMax     uint64 `json:"cpuMillicoreUsageMax"`     // @bingen:field[version=1]
-	RAMByteUsageAverage      uint64 `json:"ramByteUsageAverage"`      // @bingen:field[version=1]
-	RAMByteUsageMax          uint64 `json:"ramByteUsageMax"`          // @bingen:field[version=1]
+	PublicIPSeconds uint64 `json:"publicIpSeconds"` // @bingen:field[version=1]
```
not too sure about this one. It implies that we have a means of tracking the number of seconds that each public IP was attached to the node

ameijer replied:
I think we do have such a means?

aaliomer-ibm replied:
I'm honestly not confident about the networking metrics. I'm thinking about just taking it out completely until we talk about it and consult matt bolt at another time on a different PR. @ameijer 
Generally these are the fields we want but no idea how to get them
```
	CrossZoneBytes uint64
	CrossRegionBytes uint64
	InternetBytes uint64
	NatGatewayBytes uint64
}
```

ameijer replied:
can't we use this method to obtain this? https://github.com/opencost/opencost/blob/76a4bffbe455539bcfac26e567e84fa47d80de94/core/pkg/clustercache/clustercache.go#L186

Sean-Holcomb commented on core/pkg/model/kubemodel/namespace.go:
```diff
@@ -1,14 +1,10 @@
 	Name        string            `json:"name"`        // @bingen:field[version=1]
 	Labels      map[string]string `json:"labels"`      // @bingen:field[version=1]
 	Annotations map[string]string `json:"annotations"` // @bingen:field[version=1]
-	Start       time.Time         `json:"start"`       // @bingen:field[version=1]
```
In general we are using stat and end on all types. start and end represent the time at which we started receiving metrics for a particular object

ameijer replied:
agreed, we should keep these

Sean-Holcomb commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
+	Agent        string   `json:"agent"`        // @bingen:field[version=1]
+	AgentVersion string   `json:"agentVersion"` // @bingen:field[version=1]
+	K8sVersion   string   `json:"version"`      // @bingen:field[version=1]
+	ClusterType  string   `json:"type"`         // @bingen:field[version=1]
```
What is cluster type?

aaliomer-ibm replied:
openshift, k8s, etc

aaliomer-ibm replied:
gonna go away

ameijer replied:
looks like it is still present. @Sean-Holcomb are you good with this getting tossed?

Sean-Holcomb commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
+	Provider     Provider `json:"provider"`     // @bingen:field[version=1]
+	Account      string   `json:"account"`      // @bingen:field[version=1]
+	Name         string   `json:"name"`         // @bingen:field[version=1]
+	Agent        string   `json:"agent"`        // @bingen:field[version=1]
```
What is Agent?

aaliomer-ibm replied:
agent type. I'm actually reverting this file. I like your suggestion of the heartbeat mechanism better

ameijer replied:
looks to be still present

Sean-Holcomb commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
+	Account      string   `json:"account"`      // @bingen:field[version=1]
+	Name         string   `json:"name"`         // @bingen:field[version=1]
+	Agent        string   `json:"agent"`        // @bingen:field[version=1]
+	AgentVersion string   `json:"agentVersion"` // @bingen:field[version=1]
```
This code is in Opencost, it doesn't not have a notion of any parent application that is calling it. The bingen codex do have a version, maybe we could include that here.

Sean-Holcomb commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,17 @@
+	AgentVersion string   `json:"agentVersion"` // @bingen:field[version=1]
+	K8sVersion   string   `json:"version"`      // @bingen:field[version=1]
+	ClusterType  string   `json:"type"`         // @bingen:field[version=1]
+	Region       string   `json:"region"`       // @bingen:field[version=1]
```
I know that generally speaking a cluster will always be in a single region, but it is possible to create a cluster that spans multiple regions.

ameijer replied:
looks to be still present

Sean-Holcomb commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,118 @@
+	PowerWattSeconds float64 `json:"powerWattSeconds"` // @bingen:field[version=1] GPU device power consumption in watt-seconds (Joules)
+	PowerWattMax     float64 `json:"powerWattMax"`     // @bingen:field[version=1] GPU device max power consumption in watts
+	// Version 2 fields - Lifecycle tracking
+	Start           *time.Time `json:"start,omitempty"` // @bingen:field[version=1] - Device availability start
```
consistence on start and end type

Sean-Holcomb commented on core/pkg/model/kubemodel/pvc.go:
```diff
@@ -0,0 +1,38 @@
+	VolumeAttributes map[string]string `json:"volumeAttributes,omitempty"` // @bingen:field[version=1]
+	// PVC lifecycle timestamps
+	Start time.Time  `json:"start"`         // @bingen:field[version=1] - PVC creation timestamp
+	End   *time.Time `json:"end,omitempty"` // @bingen:field[version=1] - PVC deletion timestamp (nil if still active)
```
end consistence

Sean-Holcomb commented on core/pkg/model/kubemodel/window.go:
```diff
@@ -6,4 +7,6 @@ import "time"
 	Start time.Time `json:"start"` // @bingen:field[version=1]
 	End   time.Time `json:"end"`   // @bingen:field[version=1]
+	// Sometimes the duration is smaller than the time between start and end (missing data for example)
+	DurationSeconds uint64 `json:"durationSeconds"` // @bingen:field[version=1]
```
Can you elaborate a bit more?

aaliomer-ibm replied:
i can add an example to the comment :
`if the agent is down for 20 minutes in the middle of an hour span, then start and end will be an hour span but the duration will be only 40 minutes to account for the missing 20 minute metric window`

Sean-Holcomb replied:
Window describes a period of time delimited by a start and end in which the runtimes of various entites can occur. The uptime of the agent is inconsequential to the window.

Sean-Holcomb commented on core/pkg/model/kubemodel/namespace.go:
```diff
@@ -1,14 +1,10 @@
 // @bingen:generate:Namespace
 type Namespace struct {
 	UID         string            `json:"uid"`         // @bingen:field[version=1]
-	ClusterUID  string            `json:"clusterUID"`  // @bingen:field[version=1]
```
Good call

ameijer replied:
why would having a namespace point to a cluster be a bad thing?

aaliomer-ibm replied:
its one cluster for the entire kubemodelset

ameijer replied:
oh! of course

Sean-Holcomb reviewed (changes requested):
Just looking for elaboration with most of the comments, and some standardization with some of the fields

ameijer commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+// @bingen:generate:Device
+// Device represents a physical GPU device with DCGM integration (provisioned resource)
```
I'd remove all gpu references in this comment if we are trying to generalize

ameijer commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,118 @@
+	// Version 2 fields - Lifecycle tracking
+	Start           *time.Time `json:"start,omitempty"` // @bingen:field[version=1] - Device availability start
+	End             *time.Time `json:"end,omitempty"`   // @bingen:field[version=1] - Device availability end
+	DurationSeconds uint64     `json:"durationSeconds"` // @bingen:field[version=1] - Duration device was available
```
I don't think we need duration w/ start and end

aaliomer-ibm replied:
you're right. downtime is not relevant in this case probably

aaliomer-ibm replied:
actually we do. we have MemoryKiBSeconds and other cumulative units. anywhere we have those, we need duration

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -1,20 +1,26 @@
+	NetworkInternetEgressBytes uint64 `json:"networkInternetEgressBytes,omitempty"` // @bingen:field[version=1]
+	NetworkCrossRegionBytes    uint64 `json:"networkCrossRegionBytes,omitempty"`    // @bingen:field[version=1]
+	NetworkSameRegionBytes     uint64 `json:"networkSameRegionBytes,omitempty"`     // @bingen:field[version=1]
+	NetworkIntraAZBytes        uint64 `json:"networkIntraAzBytes,omitempty"`        // @bingen:field[version=1]
```
should we break this out to a new type? check out the networkInsights objects in this repo

aaliomer-ibm replied:
removed for now till we think it properly

ameijer commented on core/pkg/model/kubemodel/pv.go:
```diff
@@ -0,0 +1,46 @@
+	// Cloud provider's volume identifier
+	ProviderVolumeID string `json:"providerVolumeId,omitempty"` // @bingen:field[version=1]
+	// ReadWriteOnce, ReadWriteMany, ReadOnlyMany
+	AccessModes []string `json:"accessModes,omitempty"` // @bingen:field[version=1]
```
do we need this?

aaliomer-ibm replied:
Brock think it is useful. better have it and not need it

ameijer commented on core/pkg/model/kubemodel/service.go:
```diff
@@ -36,4 +58,15 @@ type Service struct {
+	NetworkInternetEgressBytes uint64 `json:"networkInternetEgressBytes,omitempty"` // @bingen:field[version=1]
+	NetworkCrossRegionBytes    uint64 `json:"networkCrossRegionBytes,omitempty"`    // @bingen:field[version=1]
+	NetworkSameRegionBytes     uint64 `json:"networkSameRegionBytes,omitempty"`     // @bingen:field[version=1]
+	NetworkIntraAZBytes        uint64 `json:"networkIntraAzBytes,omitempty"`        // @bingen:field[version=1]
```
services don't have network traffic, they are just an abstraction

aaliomer-ibm replied:
should i remove 59, 60?

ameijer replied:
probably, I think those would then be computed from other network data

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,23 @@
+//nolint:stylecheck
```
can we remove the linter disables?

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -1,23 +1,95 @@
-	StorageByteUsageMax                 uint64    `json:"storageByteUsageMax"`                 // @bingen:field[version=1]
+	PodUID                     uuid.UUID            `json:"podUid"`                              // @bingen:field[version=1]
+	Name                       string               `json:"name"`                                // @bingen:field[version=1]
+	DurationSeconds            uint64               `json:"durationSeconds"`                     // @bingen:field[version=1]
```
couldn't duration be computed as end - start time?

aaliomer-ibm replied:
no because if there is the agent goes down for any reason then it wont be accurate. remember that the model needs to be mergable

Sean-Holcomb replied:
It is still unclear to me why Kubemodel needs to be mergable.

aaliomer-ibm replied:
1. In Cldy, we had a lot of business logic that is static window aware (daily, hourly, etc...). That led to a very ugly code
2. Assuming requirements wont change or assuming we'll just do a rewrite is not a good idea
3-A. in Cldy, storage was an issue in multi-tenant environment. Kubecost doesn't support multi-tenant yet and have not faced this issue yet. Cldy has. We need to be able to combine smaller windows into a larger one without losing information. hence my push for units like seconds, byteSeconds, etc. it is something cldy didn't do and kubecost did which is brilliant on your side. 
3-B. Most containers workloads dont change drastically. If you collect data every 10 minutes will be ~5-6 times bigger than once an hour
5.  You have to assume interruptions with the agent. Also you have to assume non-contiguous use cases. For example getting daytime cost consumption 
6. Saying that mergable units (like seconds, byteseconds, bytes, etc) is a downstream concern is just not true. Those are what we call contagious units. It will impact what use-cases you're able to support downstream
7. You need flexibility. we don't know the future nor the use cases. and i don't think there is a cost for it to do it this way


Those are some of the reasons off the top of my head. I reduced those points into two requirements:
1. Dynamic windows
2. Mergable units

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,13 +1,23 @@
+	ClusterType  string    `json:"type"`            // @bingen:field[version=1]
+	Region       string    `json:"region"`          // @bingen:field[version=1]
+	Start        time.Time `json:"start,omitempty"` // @bingen:field[version=1]
+	End          time.Time `json:"end,omitempty"`   // @bingen:field[version=1]
```
set all bingen fields to 'ignore' until we finish hydrating them

ameijer commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,79 @@
+	ModelName        string    `json:"modelName"`        // @bingen:field[version=1]
+	IsShared         bool      `json:"isShared"`         // @bingen:field[version=1] GPU sharing information
+	SharePercentage  float64   `json:"sharePercentage"`  // @bingen:field[version=1]
+	UsageSeconds     float64   `json:"usageSeconds"`     // @bingen:field[version=1] GPU seconds available
```
GPU references in comments

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -1,19 +1,22 @@
+	Labels               map[string]string `json:"labels,omitempty"`      // @bingen:field[version=1]
+	Annotations          map[string]string `json:"annotations,omitempty"` // @bingen:field[version=1]
+	DurationSeconds      uint64            `json:"durationSeconds"`       // @bingen:field[version=1]
+	NetworkTransferBytes uint64            `json:"networkTransferBytes"`  // @bingen:field[version=1]
```
```suggestion
	NetworkTransferBytes []NetworkUsage          `json:"networkTransferBytes"`  // @bingen:field[version=1]
```
```
type NetworkUsage struct {
    Service          string
	Bytes            float64                 `json:"bytes"`
	EndPoint         string                  `json:"endPoint"`
	TrafficDirection NetworkTrafficDirection `json:"trafficDirection"`
	TrafficType      NetworkTrafficType      `json:"trafficType"`
}
```

ameijer replied:
```
type NetworkService struct {
    ServiceName string 
    TotalBytes float64/uint64
}

type NetworkTraffic struct {
  CrossZoneBytes float64/uint64
  CrossRegionBytes float64/uint64
  InternetBytes float64/uint64
  NatGatewayBytes float64/uint64
   Services []*NetworkService
}

On Pod/Container:
{ 
    ...
    NetworkEgress *NetworkTraffic 
    NetworkIngress *NetworkTraffic
    ...
}
```

ameijer commented:
@aaliomer-ibm can you resolve the merge issues? then I will take a review pass

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -1,23 +1,94 @@
+	RAMKiBRequestSeconds       uint64               `json:"ramKiBRequestSeconds"`                // @bingen:field[version=1]
+	VolumeStorageKiBSeconds    map[uuid.UUID]uint64 `json:"volumeStorageKiBSeconds,omitempty"`   // @bingen:field[version=1]
+	VolumeStorageByteUsageMax  map[uuid.UUID]uint64 `json:"volumeStorageByteUsageMax,omitempty"` // @bingen:field[version=1]
+	CpuMillicoreLimitSeconds   uint64               `json:"cpuMillicoreLimitSeconds,omitempty"`  // @bingen:field[version=1]
```
change all to be ignore except for the ones that were not marked ignored

Sean-Holcomb replied:
In general I think we should have a policy here that fields should be marked ignored until they have code which populates them. I am fine with asperational fields to enable functionality we want, but keeping them ignored protects us from bloat if it turns out that we cannot populate those fields. Once a version goes live we are stuck with those fields.

ameijer commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,75 @@
+	// Version 2 fields - Lifecycle tracking
+	Start           time.Time `json:"start,omitempty"` // @bingen:field[version=1] - Device availability start
+	End             time.Time `json:"end,omitempty"`   // @bingen:field[version=1] - Device availability end
+	DurationSeconds uint64    `json:"durationSeconds"` // @bingen:field[version=1] - Duration device was available
```
bingen ignore

ameijer commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,75 @@
+	IsShared         bool      `json:"isShared"`         // @bingen:field[version=1] Device sharing information
+	SharePercentage  float64   `json:"sharePercentage"`  // @bingen:field[version=1]
+	UsageSeconds     float64   `json:"usageSeconds"`     // @bingen:field[version=1] Device seconds available
+	MemoryKiBSeconds uint64    `json:"memoryKiBSeconds"` // @bingen:field[version=1] Device memory capacity in KiB-seconds
```
floats, not uints

ameijer reviewed (commented):
can we also add the networking fields?

Sean-Holcomb commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -1,10 +1,14 @@
 // @bingen:generate:Cluster
 type Cluster struct {
-	UID      string    `json:"uid"`      // @bingen:field[version=1]
+	UID      uuid.UUID `json:"uid"`      // @bingen:field[version=1]
```
While I agree with the `uuid.UUID` type in principle. The ship has kind of sailed on this field being a `string` in version 1 and probably in general given that it currently exists in the wild for a paying costumer. The fact that it is true here and with names space and resourcequota, makes me feel that the whole idea should probably be abondon.

aaliomer-ibm replied:
thats a bummer. sounds good

Sean-Holcomb commented on core/pkg/model/kubemodel/namespace.go:
```diff
@@ -1,14 +1,17 @@
-	Annotations map[string]string `json:"annotations"` // @bingen:field[version=1]
-	Start       time.Time         `json:"start"`       // @bingen:field[version=1]
-	End         time.Time         `json:"end"`         // @bingen:field[version=1]
+	UID         uuid.UUID         `json:"uid"`             // @bingen:field[version=1]
```
see other UUID comment

Sean-Holcomb commented on core/pkg/model/kubemodel/resourcequota.go:
```diff
@@ -0,0 +1,40 @@
+
+// @bingen:generate:ResourceQuota
+type ResourceQuota struct {
+	UID          uuid.UUID            `json:"uid"`             // @bingen:field[version=1]
```
see previous uuid comment

Sean-Holcomb commented on core/pkg/model/kubemodel/kubemodel_test.go:
```diff
@@ -1,237 +0,0 @@
-package kubemodel
```
Why are these tests removed?

Sean-Holcomb commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -1,23 +1,94 @@
+	End                        time.Time            `json:"end"`                                 // @bingen:field[version=1]
+}
+
+func (c *Container) CpuMillicoreUsageAverage() uint64 {
```
I think this function is emblematic of a core misunderstand of Kubemodel. It is populated via existing datasources which are used to aggregate values such as these. We do not have access to a counter like CpuMillicoreSeconds rather the value is created in a sample based approach that is not susceptible to the issue that duration seconds is trying to solve.

nikovacevic commented:
@aaliomer-ibm -- I'd like to propose we use `float64` everywhere we've previously decided to use `unit64`. Here's a full analysis of the difference between the types, and their fitness for your purposes: https://apptio.atlassian.net/wiki/spaces/Kubecost/pages/934248701/KubeModel+numeric+precision

Let me know what you think.

Sean-Holcomb commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -6,22 +6,89 @@ import (
+		return 0
+	}
+	totalByteSeconds := c.TotalStorageByteSeconds()
+	return KiBToBytes(totalByteSeconds) / c.DurationSeconds
```
Do we still need this KiB conversion func

aaliomer-ibm replied:
Nope! should take it off. Good catch!

Sean-Holcomb commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+// @bingen:generate:Device
+type Device struct {
```
Are devices relevant on the CLDY side of things?

aaliomer-ibm replied:
GPU, TPU thats it for now. Looked into special networking devices that some companies use but we don't support it yet. they are usually CAPEX spend so no customer pushed for it yet

Sean-Holcomb commented on core/pkg/model/kubemodel/device.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+// @bingen:generate:Device
+type Device struct {
```
Part of the data here is coming from the DCGM exporter which specifically pertains to Nvidea GPUs I wonder I am starting to think we may want a gpu specific type that has these additional fields on it and references a specific device

aaliomer-ibm replied:
So if the node enables MIG mode (virtual GPU), then we'll need to support the concept of a virtual device. The device concept itself is a first class k8s concept.
Maybe f `virtual` flag with `physicalDeviceUID` field? 
But to answer your original question, i think device is fine

Sean-Holcomb commented on core/pkg/model/kubemodel/window.go:
```diff
@@ -6,4 +6,8 @@ import "time"
+	// Sometimes the duration is smaller than the time between start and end (missing data for example)
+	// If the agent is down for 20 minutes in the middle of an hour span, then start and end will be
+	// an hour span but the duration will be only 40 minutes to account for the missing 20 minute metric window
+	DurationSeconds Measurement `json:"durationSeconds"` // @bingen:field[version=1]
```
This still doesn't seem to fit to me. It definitely cannot be bingen version 1

aaliomer-ibm replied:
Ah should've been `ignore`
The idea is that is the duration for how long the agent was running between the start and end times. If the agent was down for 10 minutes but start/end is an hour, then it would be 50 minutes. 
I think it needs a different name maybe a different place as well so i'll remove it. 
Not sure where it should be though

Sean-Holcomb reviewed (commented):
The main concern I have with any changes here is that they need to be hydrated and we don't currently have the metric querier functions to do so. As long as we leave things on bingen ignore until we have that capability I think it is fine to move forward with this version of the model.

Besides that there are still a lot of references to KiB which should probably go.

aaliomer-ibm committed:
Add kubemodel implementation with device support and resource management

This commit consolidates all kubemodel changes including:
- Device and device usage tracking
- PV/PVC support
- Enhanced container, pod, node, and namespace models
- Statistics and merge functionality
- Diagnostic improvements
- Remove all KIB references

Co-authored-by: Alex Meijer <ameijer@users.noreply.github.com>
Co-authored-by: Hunter Metcalfe <106991365+HMetcalfeW@users.noreply.github.com>
Co-authored-by: Kaelan Patel <32113845+kaelanspatel@users.noreply.github.com>
Co-authored-by: Matt Bolt <mbolt35@gmail.com>
Co-authored-by: Nick Curie <32180999+nickcurie@users.noreply.github.com>
Signed-off-by: aaliomer-ibm <aaliomer@ibm.com>
